File: 209.minimum-size-subarray-sum.java
/*
 * @lc app=leetcode id=209 lang=java
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int left = 0, right = 0;
        int sum = 0;
        int resLeft = -1, resRight = -1;
        int minLen = Integer.MAX_VALUE;
        while (right < nums.length) {
            sum += nums[right];
            // store the current shortest subarray
            if (sum >= target && minLen > right - left + 1) {
                minLen = right - left + 1;
                resLeft = left;
                resRight = right;
            }

            while (sum > target) {
                sum -= nums[left];
                left++;
                if (sum >= target && minLen > right - left + 1) {
                    minLen = right - left + 1;
                    resLeft = left;
                    resRight = right;
                }
            }

            right++;
        }

        return minLen == Integer.MAX_VALUE ? 0 : minLen;
    }
}
// @lc code=end

File: 209.minimum-size-subarray-sum.py
#
# @lc app=leetcode id=209 lang=python3
#
# [209] Minimum Size Subarray Sum
#

# @lc code=start

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        if sum(nums)<target:
            return 0

        n = len(nums)
        left, right = 0, 0
        curr_sum = 0
        res = n
        for right in range(n):
            curr_sum+=nums[right]
            while curr_sum>=target and left<=right:
                res = min(res, right-left +1)
                curr_sum-=nums[left]
                left+=1

        return res

# @lc code=end

